libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2027,6 +2027,30 @@ void TypeChecker::typeCheckABIEncodeFunctions(
 	}
 }
 
+void TypeChecker::typeCheckBytesConcatFunction(
+	FunctionCall const& _functionCall,
+	FunctionTypePointer _functionType
+)
+{
+	solAssert(!!_functionType, "");
+	solAssert(_functionType->kind() == FunctionType::Kind::BytesConcat, "");
+	solAssert(_functionCall.names().empty(), "");
+
+	typeCheckFunctionGeneralChecks(_functionCall, _functionType);
+
+	for (shared_ptr<Expression const> const& argument: _functionCall.arguments())
+		if (
+			Type const* argumentType = type(*argument);
+			!argumentType->isImplicitlyConvertibleTo(*TypeProvider::fixedBytes(32)) &&
+			!argumentType->isImplicitlyConvertibleTo(*TypeProvider::bytesMemory())
+		)
+			m_errorReporter.typeError(
+				8015_error,
+				argument->location(),
+				"bytes.concat argument can only be of type bytes memory or fixed bytes."
```

Comment:
I don't like the error message.
Maybe something like `bytes.concat argument must be implicitly convertible to bytes memory or fixed bytes.`

Reply:
I changed to slightly better option, but still open to suggestions...